Add missing developer documentation for the pre-commit hook setup [AAP-38768]

art-tapin opened the pull request:
[AAP-38768]
Original PR: [add pre-commit with ruff #42](https://github.com/ansible/metrics-utility/pull/42)

## Description

**What is being changed?**
- Added developer-focused documentation detailing how to install `uv` and `ruff` and configure the pre-commit hook.
- Expanded the `/docs` directory and updated the `README.md` to guide new contributors on setting up their local environments.

**Why is this change needed?**
- The pre-commit hook with `uv` and `ruff` has already been implemented, but there was insufficient documentation explaining how new developers should install and use these tools.
- Without clear setup instructions, contributors might struggle to replicate the required environment, leading to confusion and inconsistent code formatting.

**How does this change address the issue?**
- By providing a dedicated `/docs/developer_setup.md` file (and references in `README.md`), we ensure that new contributors can quickly and accurately configure their environment.
- This documentation closes the knowledge gap around `uv` and `ruff` usage, preventing onboarding challenges.

**Supporting Documentation:**
- [Ruff Documentation](https://docs.astral.sh/ruff/)
- [UV Documentation](https://docs.astral.sh/uv/)

---

## Testing

Since these changes are primarily documentation updates, there are no code changes to test directly. However, you can follow the new instructions to confirm they are accurate and comprehensive. First please check the PR https://github.com/ansible/metrics-utility/pull/42 introducing the changes for more clarity and then simply follow the same steps:

1. Clone the `metrics-utility` repository.
2. Install `uv` using `pip install uv`.
3. Run `uv sync` to install all dependencies.
4. Verify that `ruff`, `pytest`, `django-admin`, and `pre-commit` are installed.
5. Install the pre-commit hooks with `pre-commit install`.
6. Create a simple Python file with a linting error (e.g., an unused import) and attempt to commit. The pre-commit hook should block the commit.

**Expected Results:**
- New contributors should be able to follow the documentation to set up the environment and verify that the pre-commit hook blocks commits with linting issues.

---

## Required Actions

- [x] **Requires documentation updates**  
  - `/docs/developer_setup.md` created or updated  
  - `README.md` updated to reference developer documentation

- [ ] Requires downstream repository changes *(None)*
- [ ] Requires infrastructure/deployment changes *(None)*
- [ ] Requires coordination with other teams *(None)*

---

## Self-Review Checklist

- [x] Documentation is clear, accurate, and accessible.
- [x] No major workflow steps are missing.
- [x] All references and links are working.
- [ ] Any relevant diagrams or screenshots are included (N/A if none).
- [ ] Changes reviewed by at least one other team member.

---

## Notes for Reviewers

- These changes **do not** introduce new code or hook logic—the pre-commit hook was previously merged. (See the original PR https://github.com/ansible/metrics-utility/pull/42)
- This PR **adds and refines documentation** to fill the gap for new developers who need clear instructions on using `uv` and `ruff`.
- Future updates may migrate some of this content to the Ansible Handbook, but for now, it remains in the repo for immediate access.

ShaiahWren commented on docs/developer_setup.md:
```diff
@@ -0,0 +1,128 @@
+
+## 4. Configuring Pre-commit Hooks
+
+1. **Install Pre-commit Hooks**:
```
This is looking good. Note that the uv.lock file is already part of the project and pre-commit _should_ already be installed with `uv sync`, but these are good instructions from a 'starting from scratch' point of view. Wdyt @himdel ?

himdel replied:
`pre-commit install` isn't actually about installing the package, but about registering it with git, so this is still needed

art-tapin replied:
I'll mark it as resolved then :+1: 

Thank you @ShaiahWren and @himdel

himdel commented:
The changes look good :+1: 

Just, wrong repo ... we already have developer docs in the [handbook](https://github.com/ansible/handbook/tree/main/The%20Ansible%20Engineering%20Handbook/docs/AAP/Services/Metrics) so better put this there as well... there seems to be little point in splitting them into 2 different places and forcing people to search.

art-tapin commented:
@himdel 

Thanks for the feedback! Since metrics-utility is an open repository, I was aiming to keep the documentation open source and accessible for external contributors. The Ansible Handbook is private, which means developers outside our organization can’t easily access it, so placing this content here ensures broader visibility. If there’s no specific company policy about where developer docs must live, we’d like to maintain an open resource for anyone who wants to contribute.

If you have any other ideas or examples on how to handle this in a way that keeps content accessible while meeting internal guidelines, please let me know. I’m definitely open to discussing alternative approaches!

Thanks again!

art-tapin commented:
@Ladas @SherinV Please take a look at our recent conversation with @himdel . I would appreciate some clarity on where to store dev docs.

art-tapin commented:
@himdel Thanks again for the feedback. We had clarifications from @Ladas that documentation closely related to the upstream repo should be included in the repo itself. Internal documentation that could affect Red Hat or downstream, especially anything discussing internal deployment processes, should go in our section of the private Ansible Handbook instead.

Since this developer setup guide is specific to the open source metrics-utility repo and relevant for all contributors, including external ones, we'll keep it in the /docs folder here to ensure it's accessible.

himdel commented on docs/developer_setup.md:
```diff
@@ -0,0 +1,128 @@
+
+## 2. Prerequisites
+
+- **Python 3.8+** installed on your system.
```
3.8 might be too old, 3.11 is what awx uses

art-tapin replied:
Changed to 3.11 :saluting_face:

himdel reviewed (approved):
Looks good :+1: 

Maybe make sure this is linked from our handbook too?, so that people will find it even if they start looking there.

art-tapin commented:
Forced push to rebase and align with the recent changes in `devel`

art-tapin committed:
docs: add developer_setup.md and update README to reference /docs

- Created /docs/developer_setup.md with step-by-step instructions for setting up the development environment, installing uv and ruff, and configuring pre-commit hooks.
- Updated README.md to include links to the new documentation in the /docs directory.
- Clarified upstream documentation and mention internal Ansible Handbook

ShaiahWren reviewed (approved):
LGTM
